packages/mobile/src/web3/contracts.ts

```diff
@@ -2,25 +2,61 @@ import { newKitFromWeb3 } from '@celo/contractkit'
 import { privateKeyToAddress } from '@celo/utils/src/address'
 import { Platform } from 'react-native'
 import * as net from 'react-native-tcp'
+import { select, take } from 'redux-saga/effects'
+import sleep from 'sleep-promise'
 import { DEFAULT_FORNO_URL } from 'src/config'
 import { IPC_PATH } from 'src/geth/geth'
-import networkConfig from 'src/geth/networkConfig'
+import { store } from 'src/redux/store'
 import Logger from 'src/utils/Logger'
+import { Actions } from 'src/web3/actions'
+import { contractKitReadySelector, fornoSelector } from 'src/web3/selectors'
 import Web3 from 'web3'
 import { provider } from 'web3-core'
 
-// Logging tag
 const tag = 'web3/contracts'
 
-const web3: Web3 = getWeb3()
-let contractKit = newKitFromWeb3(web3)
+export const web3ForUtils: Web3 = new Web3() // Web3 with no provider
+const contractKitForno = newKitFromWeb3(getWeb3(true))
+const contractKitGeth = newKitFromWeb3(getWeb3(false))
 
-export function getContractKit() {
-  return contractKit
+function getWeb3(fornoMode: boolean): Web3 {
+  Logger.info(
+    `${tag}@getWeb3`,
+    `Initializing web3, platform: ${Platform.OS}, forno mode: ${fornoMode}, provider: ${
+      fornoMode ? DEFAULT_FORNO_URL : 'geth'
+    }`
+  )
+  return fornoMode ? new Web3(getHttpProvider(DEFAULT_FORNO_URL)) : new Web3(getIpcProvider())
+}
+
+export async function getContractKitOutsideGenerator() {
+  // Poll store until rehydrated
+  while (!store) {
+    Logger.debug(`getContractKitOutsideGenerator`, `Still waiting for store...`)
+    await sleep(250) // Every 0.25 seconds
+  }
+  // Use store to ensure ContractKit is ready
+  while (!contractKitReadySelector(store.getState())) {
+    Logger.debug(`getContractKitOutsideGenerator`, `ContractKit is still locked...`)
+    await sleep(250) // Every 0.25 seconds
+  }
+
+  Logger.debug(`${tag}@getContractKitOutsideGenerator`, 'ContractKit ready, returning')
+  return getContractKitBasedOnFornoInStore()
+}
+
+export function* getContractKit() {
```

Comment:
Can you clarify why CK has to wait for that action? As you've already seen, we call this *a lot*, and if we can somehow avoid sagafying this it would be much cleaner.

Reply:
CK has to wait for that action only if there is a lock on contractKit (contractKitReady = false), as it must wait for it to be unlocked. 
There is a lock on contractKit when:
1. the app was just opened and store is not yet persisted, so geth vs forno mode is not known and we must wait until it is known to start geth or forno
2. the app is in the middle of switching providers (turning on or off forno mode)